### ISSUE TITLE
[jax2tf] NotImplementedError: Call to scatter add cannot be converted with enable_xla=False

### ISSUE THREAD

**Issue by josephrocca** [google/jax#9659]
Conversions for XlaScatter [are currently unsupported](https://github.com/google/jax/blob/main/jax/experimental/jax2tf/g3doc/no_xla_limitations.md#summary-table) when using `enabled_xla=False`. I'm wondering if support could be added?

Here's the full error that I'm seeing:
```
NotImplementedError                       Traceback (most recent call last)
/usr/local/lib/python3.7/dist-packages/jax/experimental/jax2tf/impl_no_xla.py in op(*arg, **kwargs)
     52 
     53   def op(*arg, **kwargs):
---> 54     raise _xla_disabled_error(name)
     55 
     56   return op

NotImplementedError: Call to scatter add cannot be converted with enable_xla=False.
```

Here is some code that reproduces this error:
```
!pip install --upgrade flax
!pip install git+https://github.com/josephrocca/transformers.git@patch-2
```
```python
import jax
from jax.experimental import jax2tf
from jax import numpy as jnp

import numpy as np
import tensorflow as tf

from transformers import FlaxCLIPModel

clip = FlaxCLIPModel.from_pretrained("openai/clip-vit-base-patch32")

def score(pixel_values, input_ids, attention_mask):
    pixel_values = jax.image.resize(pixel_values, (3, 224, 224), "nearest")
    inputs = {"pixel_values":jnp.array([pixel_values]), "input_ids":input_ids, "attention_mask":attention_mask}
    outputs = clip(**inputs)
    return outputs.logits_per_image[0][0][0]

score_tf = jax2tf.convert(jax.grad(score), enable_xla=False)

my_model = tf.Module()
my_model.f = tf.function(score_tf, autograph=False, jit_compile=True, input_signature=[
  tf.TensorSpec([3, 40, 40], tf.float32),
  tf.TensorSpec([1, 30], tf.int32),
  tf.TensorSpec([1, 30], tf.int32),
])

model_name = 'pixel_text_score_grad'
tf.saved_model.save(my_model, model_name, options=tf.saved_model.SaveOptions(experimental_custom_gradients=True))
```
Here's a public colab with that code: https://colab.research.google.com/drive/1HjpRXsa8Ue9KWiKbVVWUX6DlXoIYx2r8?usp=sharing You can click "Runtime > Run all" to see the error.

Thanks!

**Comment by josephrocca** [comment]
Note that, as documented in [this issue](https://github.com/google/jax/issues/9572), the conversion works fine if I'm just converting the `score` function. But in the above example I'm converting `jax.grad(score)`. This is the only difference.

(Aside: I'm not sure if it's common to want to convert the gradient of a function to TF, but my motivation here is that, as an initial proof of concept of getting non-trivial JAX stuff working in browsers, I'm trying to get VQGAN+CLIP tuning (or something similar) working on the web. This means I need the gradient of the image-text similarity score.)

**Comment by marcvanzee** [comment]
@josephrocca have you considered just converting the original function, and then doing the gradient computation in the TFLite code?

jax2tf supports differentiation of the converted function from Tensorflow if the flag `with_gradient=True` (which is is by default, see [here](https://github.com/google/jax/blob/main/jax/experimental/jax2tf/jax2tf.py#L192)). When you run differentiation through TF, the gradient will be computed by JAX. See [saved model and differentiation](https://github.com/google/jax/tree/main/jax/experimental/jax2tf#saved-model-and-differentiation) for more details.

**Comment by josephrocca** [comment]
@marcvanzee Hey Marc, thanks for your reply. My goal here is to create a simple pipeline to get JAX functions working in the browser, which means that I'm currently limited to 3 runtimes: [tfjs-tflite](https://github.com/tensorflow/tfjs/tree/master/tfjs-tflite), [tfjs](https://github.com/tensorflow/tfjs), and [ONNX Runtime Web](https://github.com/microsoft/onnxruntime/tree/master/js/web). There are conversion tools that can get me from JAX to any of these 3 formats. I might be wrong here, but I don't think it's possible to get any of these runtimes to give me the gradients for a pretrained model? tfjs-tflite has a subset of the functionality of tflite, but I don't think it's even possible to get the regular tflite runtime to give you gradients? That is, if I'm reading [this thread](https://discuss.tensorflow.org/t/applying-grad-cam-to-a-model-that-was-converted-to-tensorflow-lite/3810/9) correctly.

But it occurred to me just now that maybe you meant that I can compute the gradient of the function with TF, and then export that new function to tfjs-tflite/tflite/onnx? If so, do you know how that'd work (considering the code example in my original post) in terms of which TF APIs to use? I'm not familiar with TF because it's just an intermediate conversion step for me, but perhaps I'll have to learn it.

**Comment by josephrocca** [comment]
@marcvanzee I'm also getting a scatter-related error when trying to use `jax2tf` on VQGAN:
```
NotImplementedError: CLIP scatter mode not implemented in jax2tf
```
and this error when trying to use `jax2tf` on DALLE-mini (a BART-like model):
```
NotImplementedError: Call to scatter cannot be converted with enable_xla=False.
```
I'm not using `jax.grad` on these models - just trying to directly convert the models themselves.

The full errors are in the last two cells of this notebook: https://colab.research.google.com/drive/1wf4JIDFL2kEawDRgrcTz2KDYXYvF_J88

**Comment by marcvanzee** [comment]
Thanks for the details @josephrocca! I'm still working on providing better support for the gather op (#9572), once I've finished that I will work on `scatter`.

**Comment by oliverdutton** [comment]
This may be useful. I recently had to for a set of simple scatters, however I found another route so closed the request. Please check first, but it did function correctly for a subset of cases.

https://github.com/google/jax/pull/9289

**Comment by josephrocca** [comment]
Hey @marcvanzee, @gnecula, just thinking about whether I should try another route to unblock myself here (specifically trying to get Google's [LiT](https://github.com/google-research/vision_transformer/tree/main#lit-models) gradients working in the browser at this point, since it seems better than CLIP for my use case), and so I was wondering if this is likely/unlikely to be fixed any time soon (e.g. within a few months, say)? I completely understand if not! Just don't want to start work on another route if this is likely to be fixed soon. Please forgive the ping! 🙏

**Comment by oliverdutton** [comment]
Do you know the line in the code where the scatter add occurs?

**Comment by gnecula** [comment]
@josephrocca Implementing the `XlaScatter` support with `enable_xla=False` means implementing the functionality of scatter using core TF ops. An example for gather is [here](https://github.com/google/jax/blob/68dbe74225af951762934f84a7e026cb8aa173da/jax/experimental/jax2tf/impl_no_xla.py#L702). Both scatter and gather are complicated, so the strategy is to implement the most common use cases first. 

Would you be interested in contributing some support for scatter?

**Comment by josephrocca** [comment]
@oliverdutton I believe the scatter add is generated by `jax.grad`. If you're curious, [here's a notebook](https://colab.research.google.com/drive/1d0ll1yniQxDavGPFB31LVz49MDGO62o8) that minimally replicates the error for the LiT model. You can just click `Runtime > Run all` to generate the error (or just scroll down and read the error that I've saved saved from a previous run).

One workaround that I thought might work for people trying to get jax2tf working for gradients of models is to generate the gradients in TensorFlow using GradientTape. So instead of this:
```py
def score_images(variables, images):
  zimg, _, _ = lit_model.apply(variables, images=images)
  return zimg.sum() # This is just a dummy score for now.

lit_variables_tf = tf.nest.map_structure(tf.Variable, lit_variables)
score_images_grad_tf = lambda images: jax2tf.convert(jax.grad(score_images), enable_xla=False, polymorphic_shapes=[None, "(b, ...)"])(lit_variables_tf, images)

my_model = tf.Module()
my_model._variables = tf.nest.flatten(lit_variables_tf)
my_model.f = tf.function(score_images_grad_tf, autograph=False, jit_compile=True, input_signature=[
  tf.TensorSpec(shape=[None, 224, 224, 3], dtype=tf.float32, name="images"),
])

model_name = 'score_images_grad'
tf.saved_model.save(my_model, model_name, options=tf.saved_model.SaveOptions(experimental_custom_gradients=True))

# Results in NotImplementedError: Call to scatter add cannot be converted with enable_xla=False.
```
you'd use something like this (i.e. `jax2tf.convert(score_images, ...)` instead of `jax2tf.convert(jax.grad(score_images), ...)` and then create `score_images_tf_grad` using `GradientTape`):
```py
def score_images(variables, images):
  zimg, _, _ = lit_model.apply(variables, images=images)
  return zimg.sum() # This is just a dummy score for now.

lit_variables_tf = tf.nest.map_structure(tf.Variable, lit_variables)
score_images_tf = lambda images: jax2tf.convert(score_images, enable_xla=False, polymorphic_shapes=[None, "(b, ...)"])(lit_variables_tf, images)

def score_images_tf_grad(images):
  with tf.GradientTape() as tape:
    score = score_images_tf(images)
  grads = tape.gradient(score, images)
  return grads

my_model = tf.Module()
my_model._variables = tf.nest.flatten(lit_variables_tf)
my_model.f = tf.function(score_images_tf_grad, autograph=False, jit_compile=True, input_signature=[
  tf.TensorSpec(shape=[None, 224, 224, 3], dtype=tf.float32, name="images"),
])

model_name = 'score_images_tf_grad'
tf.saved_model.save(my_model, model_name, options=tf.saved_model.SaveOptions(experimental_custom_gradients=True))
```
The problem with that is that is that the resulting graph seems to be missing the input node (when I look at it with Netron), and when I try to convert it to TFLite it gives this error: "ValueError: Only support at least one signature key." It's very possible that I'm making a simple mistake there, because I have basically zero experience with TensorFlow. [Here's a minimal notebook](https://colab.research.google.com/drive/1yd4Wg3bI9F4_m2qcLWxQmSxYFSF-a4vd?usp=sharing) with the above code.

> @gnecula: Would you be interested in contributing some support for scatter?

I don't think I have the technical chops to be able to do that on my own, but I would definitely be able to give it a shot if I had some dot-point guidance on what needs to be done to achieve it - a few dot points from a pro would likely save me several days. It'd be useful to know how many days this would take a pro so I can calibrate and think about if it's worth it given the time cost.

Can I assume from your comment that it's not a high priority for your team right now - i.e. support for scatter add is unlikely to be added by the team any time soon unless someone else adds it?

**Comment by josephrocca** [comment]
@gnecula It's more likely that I'm missing something here, but to summarise: The problem is that `jax2tf.convert(f)` works fine, but `jax2tf.convert(jax.grad(f))` doesn't due to lack of scatter add conversion. So my idea was to do `jax2tf.convert(f)`, and then get the gradient using TensorFlow's `GradientTape`, and then convert to SavedModel from there. Would the latter approach still require a scatter-add implementation in `jax2tf`? It didn't throw any error in [this notebook](https://colab.research.google.com/drive/1yd4Wg3bI9F4_m2qcLWxQmSxYFSF-a4vd?usp=sharing), but as mentioned above, it's for some reason missing the input node to the network when I export it to a SavedModel.

But I just now saw @oliverdutton has opened a pull request that might fix the lack of scatter add support in jax2tf 👀😮

**Comment by oliverdutton** [comment]
I have implemented a subset of xla-free scatter which covers your case.

Adding 
```python
!git clone https://github.com/oliverdutton/jax.git && cd jax && git checkout no_xla_scatter_single_dim_simple && pip install -e .
```
to your colab does now run and checking with
```python
key = jax.random.PRNGKey(0)
images = jax.random.uniform(key, (7, 224,224,3), )
jax_output = jax.grad(score_images)(lit_variables, images)

images_tf = tf.constant(images)
tf_output = score_images_grad_tf(images_tf)

jax.tree_map(lambda x,y: jnp.abs(x.numpy()-y).max(), tf_output, jax_output)
diffs = jax.tree_map(lambda x,y: jnp.abs(x.numpy()-y).max(), tf_output, jax_output)
jnp.array(jax.tree_flatten(diffs)[0]).max()
```
does match with a max abs diff 0.0001213 which seems within tolerance

This doesn't solve the serving problem which is separate from this, I'd take a look at https://blog.tensorflow.org/2021/03/a-tour-of-savedmodel-signatures.html

**Comment by oliverdutton** [comment]
@josephrocca  https://colab.research.google.com/drive/1ePrAXbnL4QhtHVCJKHC9VHZsXzngIZXR?usp=sharing has a serving signature so converts successfully to tflite. 

Notes: 
- The input shape is concrete, I set the batch dimension to 4
- The `lit_variables` have been kept as a function input even in the tf version
- To make the `lit_variables` a valid tf input they've been flattened, with names based on the params dict names

Please check this works for you, and that the tflite model is valid.

**Comment by josephrocca** [comment]
@oliverdutton This is brilliant! Thank you! I'm checking the validity of the tflite file now. It seems a bit unusual that the .tflite file is >120mb, given that (unless I'm misunderstanding), the parameters aren't saved in it. IIUC, I'll need to save the parameters as a separate file and then load them and feed them in as the first argument during inference. 120mb seems too small to contain the actual parameters, but too big to just be due to having a huge number of nodes. That said, it seems to have too many nodes for Netron to render (so I'm guessing >100k, since I've rendered ~30k nodes in it before). Might there be something unusual happening which is causing the node count to blow up? Or is this just an inherent property of the networks generated by `jax.grad`?

Also, do you know where issues should be filed to make this process easier for future users? In terms of not having to do the dot points in your previous comment, I mean. I.e. should changes ideally be made to jax2tf, or TensorFlow (or both) to make it all work "out of the box" if someone just follows the instructions in the jax2tf readme?

**Comment by oliverdutton** [comment]
File size
- Parameters shouldn't be saved in it, beyond those for config of the network
- I could believe it being huge, have a look at the difference in lengths of jax_pr between score and grad(score) with jax.make_jaxpr to get a sense for this.

To confirm, the grads here are against the 400 separate parameters of the model, not against the image. Is this what was intended?

Where to document:
- Possibly in jax2tf where tf.saved_model is described the 'trick' of flattening can be mentioned to make nested jax inputs compatible with tf. However some things like the naming of those parameters only exist because the model is the final bit of a haiku.Model, in general there won't be simple naming and params may not be a simple dict, etc. etc. Other bits like serving signatures are completely separate from jax.

**Comment by josephrocca** [comment]
@oliverdutton Thanks again for your help and investigations! I've filed an issue [here](https://github.com/tensorflow/tensorflow/issues/56075) about the unusual tflite size and they've replicated and are looking into it. I've also filed an issue [here](https://github.com/tensorflow/tfjs/issues/6403) about the tflite file not loading with `tfjs-tflite` (the web runtime for tflite).

Regarding documenting the flattening trick, is this something that could instead be handled internally by jax2tf, so the user doesn't need to do this trick? Or is that not really possible for some reason?

**Comment by oliverdutton** [comment]
The users needs to be aware that their input signature has changed to something flat, and ideally they need to name it in a way that makes sense. They just define the model and a way to pass args to it, which can be tricky when you’re using it in a number of platforms/languages. 

There could be a utility function left around though, which people are advised to use.

**Comment by josephrocca** [comment]
@oliverdutton I see, so it's just a fundamental problem of TensorFlow not allowing nested inputs. Is it also the case that it's not possible to solve this "behind the scenes" in jax2tf if we're talking about params stored in the `_variables` property of the `tf.Module`, rather than params passed as inputs to the model? It does seem kinda strange from my newbie perspective that SavedModel can't handle nested variables and inputs - I'm guessing it's more of a historical tensorflow thing rather than a technical limitation?

Also, I'm wondering about the case where the params are stored as constants in the network rather than passed in as inputs or stored as SavedModel variables? There's not enough scatter support to get Flax CLIP working ([notebook](https://colab.research.google.com/drive/1SPeT3bBBSzZpAz0kqgfzvvYCsTXIPJIb)), so I used GradientTape to get the gradient with TF instead, and I'm storing the params as constants in the network (the goal is to tune the image, like VQGAN+CLIP image generation, so the network params are constants), but that runs into the same problem of missing input nodes ([notebook](https://colab.research.google.com/drive/1Te2uHCFOdZgRT-N4VsAeIGjASK80E4im)). That seems strange because there are no variables, and the inputs aren't nested.

> There could be a utility function left around though, which people are advised to use which wraps f and inputs.

Going by [your notebook](https://github.com/google/jax/issues/9659#issuecomment-1123334306)'s code, would the most simple example code to add to the jax2tf readme be something like this?

```python
def your_fn(nested_data):
  # function that you want to convert to TensorFlow with jax2tf

nested_data = # add an input data sample here
flat_data = jax.tree_map(tf.constant, jax.tree_flatten(nested_data)[0])
flat_names = list(flatten(nested_data))
_, nested_data_structure = jax.tree_flatten(nested_data)

def your_fn_flat(flat_data):
  nested_data = jax.tree_unflatten(nested_data_structure, nested_data)
  return your_fn(nested_data)

your_fn_flat_tf = jax2tf.convert(your_fn_flat, enable_xla=False)

my_model = tf.Module()
my_model.f = tf.function(your_fn_flat_tf, autograph=False, jit_compile=True, input_signature=[
  jax.tree_map(lambda x, name: tf.TensorSpec(x.shape, x.dtype, name=name), flat_data, flat_names),
])

model_name = 'your_fn_flat'
tf.saved_model.save(my_model, model_name,  options=tf.saved_model.SaveOptions(experimental_custom_gradients=False), signatures=my_model.f.get_concrete_function())
```

I think one problem with that is in cases like this: `list(flatten({'foo':{'bar':1}, 'foo_bar':2}))`, which returns `['foo_bar']`. It's also just kind of complicated, and I'm concerned that it would be hard for many people to use even if that were put in the jax2tf readme, especially if they have multiple inputs, and perhaps also need to use `_variables`. But maybe most users of jax2tf won't be as much of a noob as me? I'm not sure.

I'm also wondering if the `signatures=my_model.f.get_concrete_function()` bit is necessary in `tf.saved_model.save`, since we've already defined `input_signature` during the `tf.function` creation? Strangely, if I add it to the CLIP notebook that I linked above, then I get this error during ``tf.saved_model.save``:
```
ValueError: Got a non-Tensor value <tf.Operation 'PartitionedCall' type=PartitionedCall> for key 'output_0' in the output of the function __inference_score_tf_grad_95495 used to generate the SavedModel signature 'serving_default'. Outputs for functions used as signatures must be a single Tensor, a sequence of Tensors, or a dictionary from string to Tensor.
```

**Comment by oliverdutton** [comment]
> @oliverdutton I see, so it's just a fundamental problem of TensorFlow not allowing nested inputs. Is it also the case that it's not possible to solve this "behind the scenes" in jax2tf if we're talking about params stored in the `_variables` property of the `tf.Module`, rather than params passed as inputs to the model? It does seem kinda strange from my newbie perspective that SavedModel can't handle nested variables and inputs - I'm guessing it's more of a historical tensorflow thing rather than a technical limitation?

It's not a fundamental problem of tensorflow, it's due to the design decision in tensorflow serving signatures. If you run jax2tf then call the fn in tensorflow with the variables still not nested it will function fine. It's saving with a serving signature that hits you.  See https://www.tensorflow.org/tfx/serving/signature_defs.

> Also, I'm wondering about the case where the params are stored as constants in the network rather than passed in as inputs or stored as SavedModel variables? There's not enough scatter support to get Flax CLIP working ([notebook](https://colab.research.google.com/drive/1SPeT3bBBSzZpAz0kqgfzvvYCsTXIPJIb)), so I used GradientTape to get the gradient with TF instead, and I'm storing the params as constants in the network (the goal is to tune the image, like VQGAN+CLIP image generation, so the network params are constants), but that runs into the same problem of missing input nodes ([notebook](https://colab.research.google.com/drive/1Te2uHCFOdZgRT-N4VsAeIGjASK80E4im)). That seems strange because there are no variables, and the inputs aren't nested.

I am currently working on a fix that stops that `scatter-add` from ever happening, this allows for using the 'conventional' tflite pathway. #10682 
```python
import functools
serving_func = functools.partial(jax.grad(score_images), lit_variables)
x_input = jnp.zeros((4,224,224,3))
converter = tf.lite.TFLiteConverter.experimental_from_jax(
    [serving_func], [[('input1', x_input)]])
tflite_model = converter.convert()
with open('score_model_grad.tflite', 'wb') as f:
  f.write(tflite_model)
```

> I think one problem with that is in cases like this: `list(flatten({'foo':{'bar':1}, 'foo_bar':2}))`, which returns `['foo_bar']`. It's also just kind of complicated, and I'm concerned that it would be hard for many people to use even if that were put in the jax2tf readme, especially if they have multiple inputs, and perhaps also need to use `_variables`. But maybe most users of jax2tf won't be as much of a noob as me? I'm not sure.
Yep, it's problematic and hacky fix.

> I'm also wondering if the `signatures=my_model.f.get_concrete_function()` bit is necessary in `tf.saved_model.save`, since we've already defined `input_signature` during the `tf.function` creation? Strangely, if I add it to the CLIP notebook that I linked above, then I get this error during `tf.saved_model.save`:
> ```
> ValueError: Got a non-Tensor value <tf.Operation 'PartitionedCall' type=PartitionedCall> for key 'output_0' in the output of the function __inference_score_tf_grad_95495 used to generate the SavedModel signature 'serving_default'. Outputs for functions used as signatures must be a single Tensor, a sequence of Tensors, or a dictionary from string to Tensor.
> ```
I have no idea what's happening there, looks like you're passing in an op rather than tensor which seems very wrong.

**Comment by oliverdutton** [comment]
Looked at CLIP, it's a great case to have, I hadn't considered when the update was scalar I'll add to the tests.

It's actually just a trivial update, [0.] -> [1.] by indexing at [0].

```python
NotImplementedError: Call to scatter add cannot be converted with enable_xla=False.

> /raid/app/oliver/repos/jax2/jax/jax/experimental/jax2tf/impl_no_xla.py(56)op()
     54 
     55   def op(*arg, **kwargs):
---> 56     raise _xla_disabled_error(name)
     57 
     58   return op

ipdb> u
> /raid/app/oliver/repos/jax2/jax/jax/experimental/jax2tf/jax2tf.py(826)invoke_impl()
    824     def invoke_impl() -> TfVal:
    825       if impl_needs_avals:
--> 826         return impl(
    827             *args_tf,
    828             _in_avals=args_avals,  # type: ignore

ipdb> args_tf
[<tf.Tensor: shape=(1,), dtype=float32, numpy=array([0.], dtype=float32)>, <tf.Tensor: shape=(1,), dtype=int32, numpy=array([0], dtype=int32)>, <tf.Tensor: shape=(), dtype=float32, numpy=1.0>]
```

**Comment by josephrocca** [comment]
@oliverdutton Thanks again for your guidance and fixes! Looking forward to those PRs landing.

@marcvanzee Thanks for your comments Marc! Replies below.

> I think we should try to understand why this happens in a simple example and see whether we can fix it, because perhaps there is a bug in jax2tf where we are not converting the dict from JAX to TF correctly. Do you have any thoughts? If not, I will file an issue for this.

I think Oliver understands this better than me, but from his comments my understanding is that there were three issues: (1) Lack of some necessary scatter support (#10653). (2) Some scatters being introduced where they weren't actually necessary (#10682). (3) SavedModel signatures not being able to be nested/PyTrees.

But I think you're right - if (3) is a problem, then why does it only occur for the `grad` of the function? I also noticed that it occurs in [this notebook](https://colab.research.google.com/drive/1Te2uHCFOdZgRT-N4VsAeIGjASK80E4im) despite there being no nesting in the inputs or outputs (and no `_variables` either - params are stored as constants in the network).

> Actually if you want to convert to TFLIte, I don't see why you would convert to SavedModel in the first place. Instead, a much easier road is to convert JAX --> TF --> TFLite directly using the TFLite converter.

🤦 I didn't realise that the TFLite converter had a `from_concrete_functions` method, thank you! I guess that skips the SavedModel signature issue. One problem though: The resulting `vit.tflite` file from the notebook you linked is only 588kb, and according to [Netron](https://netron.app/), only has only two notes - an input node and an output node. So something seems to be silently going wrong during the conversion? I did notice that there are a few (not-super-informative) warnings.

> I think this is because you want to use the [tfjs-tflite](https://github.com/tensorflow/tfjs/tree/master/tfjs-tflite) converter, right? Why? I believe that tool is still experimental, and it seems to me you have better changes using TFjs directly instead.

There are currently 3 viable runtimes for getting ML models working on the web - [ONNX Runtime Web](https://github.com/microsoft/onnxruntime/tree/master/js/web), [tfjs-tflite](https://github.com/tensorflow/tfjs/tree/master/tfjs-tflite), and [tfjs](https://github.com/tensorflow/tfjs). Unfortunately they all suffer from patchy op support (I think it's just hard for them to keep up - hopefully [WebNN](https://github.com/webmachinelearning/webnn) and/or [ModelLoader](https://github.com/webmachinelearning/model-loader) will fix this). I did try tfjs and filed a Github issue for the required op support ([1](https://github.com/tensorflow/tfjs/issues/6395)). I also tried ONNX (which usually has the best op coverage - at least in its Wasm backend), but I think that failed due to the input/output signature stuff. tfjs-tflite seems to have the required ops and with Oliver's signature flattening trick I did manage to get it to produce a model ([although it's 30x larger than it should be](https://github.com/tensorflow/tensorflow/issues/56075) - Oliver posted a theory in that thread), but I'm still waiting to hear back from the tfjs-tflite team about [an opaque initialization error](https://github.com/tensorflow/tfjs/issues/6403).

But at the moment it may be that without Oliver's "signature flattening" trick, TFLite is the only option (assuming we get `from_concrete_functions` working properly RE the 588kb model output), since, IIUC, both ONNX and tfjs require using SavedModel as an intermediary, and as discussed, JAX --> TF --> SavedModel currently produces a network without an input node due to the seemingly-nesting-related SavedModel signature problem.

> Finally, some of my colleagues recently published a [blog post](https://ai.googleblog.com/2022/04/locked-image-tuning-adding-language.html) containing a [live LiT demo](https://google-research.github.io/vision_transformer/lit/)

I did see that! I figured that thanks to that work I wouldn't have to do much work on the text side of things (though I'm not sure if the browser-ported tokenizer is open-sourced on Github yet?). AFAIK they didn't port the image model (and instead pre-computed the image embeddings for a set number of images), but luckily the LiT model [seems to convert easily](https://colab.research.google.com/drive/1tOvVBnGFbI09Ubp2VEpZKQfBGhhm52H2) - it's just the introduction of `jax.grad` that has given rise to the scatter and SavedModel signature issues.

In case you're curious, the initial goal here is to add an option to use LiT (instead of CLIP) in [this browser demo](https://josephrocca.github.io/clip-image-sorter/), and also to see how CLIP and LiT compare in gradient-based tuning/generation of images in the browser (in particular, collaborative tuning over WebRTC).

> I know this conversion path is still quite rough, but I am really glad you are trying this since it is efforts like this that help us figuring out what is unclear!

Thank you for your work on making jax2tf! I am very excited about getting JAX models working in the browser, and I'm dreaming of the day that I can just do *everything* in JAX and then port it to the browser with a conversion process that "just works". In my wildest dreams I run JAX models [directly in the browser](https://github.com/google/jax/issues/1472). There are some fearless people (like @verhulstm) over in the PyScript/Pyodide ecosystem who are working on getting PyTorch and TensorFlow working in the browser, so I'm counting that as my license to dream about JAX running in the browser 😅

**Comment by gnecula** [comment]
>Thank you for your work on making jax2tf! I am very excited about getting JAX models working in the browser, and I'm dreaming of the day that I can just do everything in JAX and then port it to the browser with a conversion process that "just works". In my wildest dreams I run JAX models https://github.com/google/jax/issues/1472.

We thank you for your interest and patience.

**Comment by oliverdutton** [comment]
Summary:
[[jax2tf] NotImplementedError: Call to scatter add cannot be converted with enable_xla=False](https://github.com/google/jax/issues/9659#) began as supporting XLA free scatters.

XLA free scatters now supported, so this ticket is theoretically closed. However the root issue was the ability to run tflite models with the grad of LiT and CLIP.

The way the scatters ended up in these models prompted me to solve an issue I've had before in #10682 so the scatter simply doesn't appear anyhow. So two things have happened: one pull request which makes scatters xla-free, and another which avoids most scatters appearing to begin with.

So two tasks:
1 - grad of LiT in tflite
2 - grad of CLIP in tflite

### LiT grad
For LiT this [notebook](https://drive.google.com/file/d/1cX2cRC5yNbuPTgGdFEEbPFJTVWgyW-Z9/view?usp=sharing) does build the tflite model (which does end up huge) and runs, matching jax output with a reasonable tolerance.

### CLIP grad
CLIP is ongoing, zero grads are observed in the tflite model while they are non zero in the jax version.




However,

**Comment by oliverdutton** [comment]
CLIP grads working correctly here [notebook](https://drive.google.com/file/d/1elQpDL1ysaOMncMnnuHEalnh4HAAeC53/view?usp=sharing), though I've used a matching image size straight away. Please experiment with it and see if the resize steps are whats causing the issues.

**Comment by marcvanzee** [comment]
@oliverdutton @josephrocca I feel that we are having multiple conversation on this issue and it is a bit difficult for me to keep track. I think we can separate the following four issues:

1. Add support for scatter when `enable_xla=False`. Main discussion in this issue.
    * ==> Suggestion: close this issue once @oliverdutton's PR is in.
2. Converting LiT to TFLite when `enable_xla=False`. 
    * ==> Suggestion: Seems to be working fine, so nothing to do. (To be confirmed by @josephrocca).
3. Converting CLIP to TFLite when `enable_xla=False`.
    * ==> Suggestion: Create a new issue for this.
4. Converting to SavedModel needs annoying parameter flattening.
    * ==> Suggestion: Create a new issue for this.

Does this match how you see it?

**Comment by marcvanzee** [comment]
Thanks @oliverdutton. Why are you hazy about 4? I first like to understand better why we would need the flattening logic (hopefully we don't need this at all!). I filed an issue #10725.

Also, why wouldn't you recommend the path that uses jax2tf?

**Comment by oliverdutton** [comment]
4. I did a quick fix which worked, but I spent little time thinking about it before moving onto `tflite.experimental_from_jax`. I think it may be just if you want it a serving function, but I don't use tensorflow serving/related things ever so I just don't know.

**Comment by josephrocca** [comment]
@oliverdutton Thanks for the CLIP grads notebook and for making me aware of the tflite converter's `experimental_from_jax`! I'm looking into this now. 

> Converting to ONNX however, means jax2tf enable_xla=False still has a specific use. (That's the one I know of, there are likely many other cases I'm not aware of)

This might be going too far off topic on an already complicated/meandering issue, but I just wanted to note that `tf2onnx` has added tflite-to-onnx conversion last year, so if tflite's `experimental_from_jax` route is quite robust, then the question of whether to direct people to `enable_xla=False` or `experimental_from_jax` + `tf2onnx` comes down to the robustness of tf2onnx's tflite conversion versus the robustness of the `jax2tf` + `enable_xla=False` route.

I've just tried `tf2onnx` ([public notebook](https://colab.research.google.com/drive/1DygMV-Nlae6BEJmZjN_laIdfYL33BbE0)) on the CLIP grads notebook, but ran into an error, which I've filed [an issue](https://github.com/onnx/tensorflow-onnx/issues/1939) in tf2onnx repo for. For now it might be handy to mention both routes in the readme?

> Converting LiT to TFLite when enable_xla=False. ==> Suggestion: Seems to be working fine, so nothing to do. (To be confirmed by josephrocca).

I didn't end up getting LiT working in the browser with tflite due to [some issues](https://github.com/tensorflow/tfjs/issues/6403) that may be specific to the tfjs-tflite browser runtime (rather than a problem with the conversion process). I also haven't tried Oliver's `experimental_from_jax` on the LiT models. I did end up getting it working via tf2onnx ([minimal demos](https://github.com/josephrocca/lit-encoder-js/), [image sorting demo](https://josephrocca.github.io/clip-image-sorter/)), but I had to avoid serializing to `SavedModel` and instead use `tf2onnx`'s Python API (`tf2onnx.convert.from_function`) on the `jax2tf` output to get it to work (I can't remember the specific error), as shown [here](https://github.com/josephrocca/lit-encoder-js/blob/main/LiT_conversion_jax2tf_to_SavedModel_to_ONNX_text_%26_image_models.ipynb). I'm not sure if there are any takeaways for jax2tf there.

**Comment by marcvanzee** [comment]
Hello all! Given that @oliverdutton's PR #10653 is merged and we now have initial support for the scatter op, I am closing this issue.

Two closing remarks:

@josephrocca 
> 🤦 I didn't realise that the TFLite converter had a `from_concrete_functions` method, thank you! I guess that skips the SavedModel signature issue. One problem though: The resulting `vit.tflite` file from the notebook you linked is only 588kb, and according to [Netron](https://netron.app/), only has only two notes - an input node and an output node. So something seems to be silently going wrong during the conversion? I did notice that there are a few (not-super-informative) warnings.

Hmm you are right. Indeed it seems that when we wrap the function in `jax.grad` it doesn't include the parameters. I think the reason is that we still have a bug in the conversion of `jax.grad`, which results in the TFLite model outputting only zeros, which is then optimized by the TFLite model by just throwing away the params (since it doesn't need them). I have filed a separate issue for this (#10819).

@oliverdutton 
> Route 2. tflite.experimental_from_jax uses the HLO itself, and doesn't save any intermediate files.
> While both are perfectly valid, the second seems like less possible pain points.

Generally I would really recommend the jax2tf --> tflite path because it has most support and is actively being worked on, so you have the biggest chance that any issues you run into will actually be addressed.

Also, note you don't have to store any intermediate files: jax2tf converts your JAX function to a TF function, which can directly be converted to a TFLite model using TFLIte's `from_concrete_function`.

**Comment by marcvanzee** [comment]
Another update: I think I was actually wrong in saying that you should always convert JAX --> TFLite through `from_concrete_function`. I think the main use case for going through a `SavedModel` is to be able to support shape polymorphism. This doesn't seem to work when using `from_concrete_function` because the signature looks quite odd, and I think it only supports a single shape. Moreover, TFLite actually recommends the SavedModel path themselves as well [here](https://www.tensorflow.org/lite/convert#convert_a_savedmodel_recommended_)!. Our current MNIST example does not show how to go from a JAX model to a TFLite model with shape polymorphism.

I have filed another issue saying we should improve this (#10821).
